### PR TITLE
[RFC] [WIP] add ESTIMATOR msgs, MAV_FIELD and MAV_SENSOR_TYPE enum

### PR DESCRIPTION
**Pull request by TSC21**
I am reopening this to have it available for discussion.

Associated issues/PRs:
https://github.com/mavlink/mavlink/pull/667
https://github.com/mavlink/mavlink/pull/873
https://github.com/mavlink/mavlink/pull/877

Note: the reported issue on https://github.com/mavlink/mavlink/pull/877 was actually a rebase problem, as `ESTIMATOR_INNOV_COV` shouldn't have got in.

@LorenzMeier, @julianoes, @priseborough, @jgoppert, @thomasgubler.

Edited: Some actions from discussion:

* [ ] rename euler body rate to just body rate (https://github.com/mavlink/mavlink/pull/880#issuecomment-451058999)

### CODE REVIEW

**Comment by TSC21** [comment]
@LorenzMeier can we open a discussion about this?

**Comment by TSC21** [comment]
@jgoppert your input here is also important

**Comment by hamishwillee** [comment]
@TSC21 If you have no joy otherwise, one thing that can be done to help move a PR along is to add the [Dev Call](https://mavlink.io/en/about/support.html#dev_call) label and attend the meeting.

**Comment by TSC21** [comment]
@hamishwillee my problem is actually be able to attend the devcall :D

**Comment by hamishwillee** [comment]
@TSC21 
- What needs to happen to progress this? 
- Does this address all the cases of https://github.com/mavlink/mavlink/pull/878? - if so can we close that as superseded by this?

**Comment by TSC21** [comment]
> - What needs to happen to progress this?

@hamishwillee this has been merged before. I think at this point what we need to address are the interfaces on the consumers side (I can spoke by the PX4 side, but maybe Ardupilot has a word to say about this as well).

> - Does this address all the cases of #878? - if so can we close that as superseded by this?

The idea behind this is that you have a generic estimator interface that allows the autopilot Firmware to pass and receive estimator data. It's agnostic, in the sense that it's not limited to the number of states and the specific estimator interfaces. #878 is an implementation that will continue the focus on the convergence of the interfaces to specific estimators, while what we are trying to achieve here is a more broad approach to the estimators interfaces.
By any chance PR replaces #878, since they have different purposes. What needs to happen, in this sense, is that whatever extension we do to `ESTIMATOR_STATUS` message is a generic approach to an estimator status message, not a specific approach to a commonly used estimator. So, given that, that extension could be matched according to what this PR offers in terms of the considered states. Then, if there are more states to consider, those can be extended on both sides.

**Comment by hamishwillee** [comment]
Hi @TSC21 

Thanks very much. In particular this was not at all obvious:

> The idea behind this is that you have a generic estimator interface that allows the autopilot Firmware to pass and receive estimator data. It's agnostic, in the sense that it's not limited to the number of states and the specific estimator interfaces. 

Why is this useful? Who is the consumer?

You talk about the Firmware receiving estimator data - is the idea you might have an offboard estimator? Is that likely?
Similarly, this can be output by firmware - who would consume it? GCS? 

Basically it would be useful to have all the things expected in the generic RFC process: https://github.com/mavlink/rfcs

That makes it much easier to engage with ArduPilot and other interested parties. At the moment I can't engage anyone because I don't see why this is considered useful.

**Comment by TSC21** [comment]
> You talk about the Firmware receiving estimator data - is the idea you might have an offboard estimator? Is that likely?

Yes that is likely. And that the same time you can have the data being sent to the ground stations or SLAM platforms that require closed-loop pose estimation with covariance to work properly. The broad situation here is to actually allow the usage of onboard or offboard estimators that are not tide to the firmware specifications but rather to an accepted common interface - in this case, it's Mavlink. Narrow the message interfacing to specific estimators results in removing flexibility to anyone trying to move forward with their own estimators. At least, from this point, we can agree on a common estimator interface.

I am out of bandwidth to move to a full spec RFC at this moment. But there may be several people working with GNC projects interested on this. @jgoppert can bring is thoughts as well if he as some time, since he was the first to bring this in.

**Comment by TSC21** [comment]
@hamishwillee the first tackle on this matter was done in https://github.com/mavlink/mavlink/pull/667. You have there a basic explanation for an example where this can become handy.

**Comment by hamishwillee** [comment]
Thanks @TSC21 . I was hoping for the full RFC because eventually we'll want to document this protocol, and the RFC would have fed directly into that. In addition, RFC processes are intended to force discipline on the design. Oh well. I guess all we can do is ping people until either they answer or we get bored.

**Comment by hamishwillee** [comment]
@auturgy @WickedShell Do you know anyone in the ArduPilot community who has an interest in [Simultaneous Localisation and Mapping (SLAM)](https://www.kudan.eu/kudan-news/an-introduction-to-slam/) and might be able to provide feedback on the proposed APIs ([discussed in #667](https://github.com/mavlink/mavlink/pull/667#issuecomment-275555572)) to share estimator data in a generic way?

 @jgoppert @priseborough You've had a couple of years to think about this now - has anything changed that would make you approach this problem differently? Is your recommendation that we push these messages is as is?

**Comment by jgoppert** [comment]
Paul suggested that we rename euler body rate to just body rate, I'm fine with that change.

**Comment by mhkabir** [comment]
@hamishwillee @jgoppert @TSC21 Interest in reviving this?

**Comment by TSC21** [comment]
@mhkabir yep

File: message_definitions/v1.0/common.xml [hunk 1]
```diff
@@ -2885,6 +2885,236 @@
         <description>North, East, Down</description>
       </entry>
     </enum>
+    <enum name="MAV_FIELD">
```
Comment (hamishwillee) [inline message_definitions/v1.0/common.xml]: MAV_FIELD_TYPE ?

**Comment by hamishwillee** [comment]
From what @jgoppert says, this should go in "as is". The only suggestion is https://github.com/mavlink/mavlink/pull/880#issuecomment-451058999

- Is there anyone who thinks that is a bad idea? 
- Any further suggestions?
- If we get this in, who is planning on using it, and when will there be an implementation?

If not, then propose we put a `<wip>` on the message and accept it pending evaluation against a real implementation.

**Comment by amilcarlucas** [comment]
ping

**Comment by hamishwillee** [comment]
Thanks @amilcarlucas ! I'll add to list to discuss in dev call tonight. BUt if no one shows interested will merge anyway.

**Comment by hamishwillee** [comment]
Discussed in the dev call last night. We decided that there is not much pull or interest in this. So closing.

If someone wants to reopen and champion this, then they can feel free, but we'd need answer to the question:

> If we get this in, who is planning on using it, and when will there be an implementation?"

**Pull request closed without merging**